File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

   class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head) return nullptr;

        // Step 1: Find the length of the list
        int len = 0;
        ListNode* temp = head;
        while (temp) {
            len++;
            temp = temp->next;
        }

        // Step 2: Check if we need to remove the head
        if (n == len) {
            ListNode* newHead = head->next;
            delete head; // Prevent memory leak
            return newHead;
        }

        // Step 3: Traverse again to find the node before the one to delete
        temp = head;
        for (int i = 0; i < len - n - 1; i++) {
            temp = temp->next;
        }

        // Step 4: Remove the target node
        ListNode* toDelete = temp->next;
        temp->next = temp->next->next;
        delete toDelete; // Free memory

        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        cur=head
        l=0
        while cur:
            l+=1
            cur=cur.next
        if n==1 and l==1:
            return
        j=l-n
        if j==0:
            head=head.next
            return head
        cur=head
        for i in range(j-1):
            cur=cur.next
        if cur.next:
            cur.next=cur.next.next
        else:
            cur.next=None
        return head

